apps/ejabberd/src/mod_last_riak.erl

```diff
@@ -0,0 +1,76 @@
+%%%----------------------------------------------------------------------
+%%% File    : mod_last.erl
+%%% Author  : Pawel Pikula <pawel.pikula@erlang-solutions.com>
+%%% Purpose : mod_last mnesia backend (XEP-0012)
+%%%
+%%%
+%%% ejabberd, Copyright (C) 2002-2014   ProcessOne
+%%% MongooseIM, Copyright (C) 2014      Erlang Solutions Ltd.
+%%%
+%%%----------------------------------------------------------------------
+
+-module(mod_last_riak).
+
+-behaviour(mod_last).
+
+-include("mod_last.hrl").
+-include("ejabberd.hrl").
+
+-define(TIMESTAMP_IDX, {integer_index, "timestamp"}).
+
+%% API
+-export([init/2,
+         get_last/2,
+         count_active_users/2,
+         set_last_info/4,
+         remove_user/2]).
+
+-spec init(ejabberd:server(), list()) ->ok.
+init(_VHost, _Opts) ->
+    %% we are using commong riak pool
+    ok.
+
+-spec get_last(ejabberd:luser(), ejabberd:lserver()) ->
+    {ok, non_neg_integer(), binary()} | {error, term()} | not_found.
+get_last(LUser, LServer) ->
+    case mongoose_riak:get(bucket_type(LServer), LUser) of
+        {ok, Obj} ->
+            Status = riakc_obj:get_value(Obj),
+            MD = riakc_obj:get_update_metadata(Obj),
+            [Timestamp] = riakc_obj:get_secondary_index(MD, ?TIMESTAMP_IDX),
+            {ok, Timestamp, Status};
+        {error, notfound} ->
+            not_found;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec count_active_users(ejabberd:lserver(), non_neg_integer()) -> non_neg_integer().
+count_active_users(LServer, TimeStamp) ->
+    Idx = {index, bucket_type(LServer), ?TIMESTAMP_IDX, TimeStamp+1, infinity()},
+    RedMF = {modfun, riak_kv_mapreduce, reduce_count_inputs},
+    Red = [{reduce, RedMF, [{reduce_phase_batch_size, 1000}], true}],
+    {ok, [{0, [Count]}]}  = mongoose_riak:mapred(Idx, Red),
```

Comment:
But to me it looks like we use the secondary index in the map reduce, don't we? Where is the bug in my reasoning? I know that in the worst case when we choose bad timestamp it may go through the entire bucket.